packages/web-frontend/cypress/scripts/generateConfig/VisualisationFilter.js

```diff
@@ -0,0 +1,93 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { keyBy, upperFirst } from 'lodash';
+
+import { getUniqueEntries, hasIntersection, toArray } from '@tupaia/utils';
+
+class VisualisationFilter {
+  constructor(db, filter) {
+    this.db = db;
+    this.filter = filter;
+  }
+
+  /**
+   * @abstract
+   */
+  // eslint-disable-next-line no-unused-vars
+  fetchRecords = async objectUrls => {
+    throw new Error('Any subclass of VisualisationFilter must implement the "fetchRecords" method');
+  };
+
+  /**
+   * @abstract
+   */
+  getBuilderKey = () => {
+    throw new Error(
+      'Any subclass of VisualisationFilter must implement the "getBuilderKey" method',
+    );
+  };
+
+  /**
+   * @protected
+   */
+  getName = () => 'visualisation';
+
+  apply = async objectUrls => {
+    if (Object.keys(this.filter).length === 0) {
+      return objectUrls;
+    }
+
+    const records = await this.fetchRecords(objectUrls);
+    const recordsById = keyBy(records, 'id');
+
+    const filterUrl = url =>
+      Object.entries(this.filter).some(([key, targetValue]) => {
```

Comment:
I used `or` logic here. I guess that `and` would be also useful, but perhaps we could change this when required?

Reply:
Actually, it wouldn't be too hard to add support for `and` - we'd just need to think a bit of how to express it in config.

I know there is existing precedent of using different syntax than the following, but one option would be:

```jsonc
   {
     // `or` logic 
     "filter": { "dataBuilder": "sum", "project": "covidau" },
     // `and` logic
     "filter": [{ "dataBuilder": "sum" }, { "project": "covidau" }]
   }
   ```
   
I'm keen on suggestions - and should probably look into existing conventions